src/mlpack/methods/ann/layer/adaptive_mean_pooling_impl.hpp

```diff
@@ -0,0 +1,122 @@
+/**
+ * @file adaptive_mean_pooling_impl.hpp
+ * @author Kartik Dutt
+ *
+ * Implementation of the Adaptive Mean Pooling layer class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_ADAPTIVE_MEAN_POOLING_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_ADAPTIVE_MEAN_POOLING_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "adaptive_mean_pooling.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+AdaptiveMeanPooling<InputDataType, OutputDataType>::AdaptiveMeanPooling()
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+AdaptiveMeanPooling<InputDataType, OutputDataType>::AdaptiveMeanPooling(
+    const size_t outputWidth,
+    const size_t outputHeight) :
+    inSize(0),
+    outSize(0),
+    inputWidth(0),
+    inputHeight(0),
+    outputWidth(outputWidth),
+    outputHeight(outputHeight),
+    reset(false),
+    deterministic(false),
+    batchSize(0)
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+AdaptiveMeanPooling<InputDataType, OutputDataType>::AdaptiveMeanPooling(
+    const std::tuple<size_t, size_t> outputShape):
+    inSize(0),
+    outSize(0),
+    inputWidth(0),
+    inputHeight(0),
+    outputWidth(std::get<0>(outputShape)),
+    outputHeight(std::get<1>(outputShape)),
+    reset(false),
+    deterministic(false),
+    batchSize(0)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void AdaptiveMeanPooling<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  batchSize = input.n_cols;
+  inSize = input.n_elem / (inputWidth * inputHeight);
+  inputTemp = arma::cube(const_cast<arma::Mat<eT>&&>(input).memptr(),
```

Comment:
`copy_aux_mem` is set to true by default, so you don't need this strictly speaking. Do you think it helps readability?

Reply:
I did this to clearly mention that we are creating a copy as most of the codebase uses same auxiliary memory. If needed, I can remove it.

Reply:
Hi @birm, I have made the changes for now, If needed I can switch back. I think this looks a lot better.

Reply:
Which one do you think looks better? I'm okay with either case personally. :)

Reply:
I think this okay for now as it is looks a bit more concise.